quickwit/quickwit-doc-mapper/src/routing_expression/mod.rs

```diff
@@ -209,26 +277,113 @@ impl Display for InnerRoutingExpr {
 enum ExprType {
     Field,
     Composite,
+    Modulo,
 }
 
-impl RoutingExpr {
-    /// Evaluates the expression applied to the given
-    /// context and returns a u64 hash.
-    ///
-    /// Obviously this function is not perfectly injective.
-    pub fn eval_hash<Ctx: RoutingExprContext>(&self, ctx: &Ctx) -> u64 {
-        if let Some(inner) = self.inner_opt.as_ref() {
-            let mut hasher: SipHasher = self.salted_hasher;
-            inner.eval_hash(ctx, &mut hasher);
-            hasher.finish()
+mod expression_dsl {
+    use nom::bytes::complete::tag;
+    use nom::character::complete::multispace0;
+    use nom::combinator::{eof, opt};
+    use nom::error::ErrorKind;
+    use nom::multi::separated_list0;
+    use nom::sequence::{delimited, tuple};
+    use nom::{AsChar, Finish, IResult, InputTakeAtPosition};
+
+    // this is a RoutingSubExpr in our DSL.
+    #[derive(Debug, PartialEq, Eq, Clone)]
+    pub(crate) enum ExpressionAst {
+        Field(String),
+        Function { name: String, args: Vec<Argument> },
+    }
+
+    #[derive(Debug, PartialEq, Eq, Clone)]
+    pub(crate) enum Argument {
+        Expression(Vec<ExpressionAst>),
+        Number(u64),
+    }
+
+    pub(crate) fn parse_expression(expr_dsl_str: &str) -> anyhow::Result<Vec<ExpressionAst>> {
+        let (i, res) = routing_expr(expr_dsl_str)
+            .finish()
+            .map_err(|e| anyhow::anyhow!("error parsing routing expression: {e}"))?;
+        t(eof(i))?;
```

Comment:
I did not know about this pattern.
Did it help the compiler?

Can you either:
- remove the pattern to something less ninja. I suspect something like this compiles.
- rename `t` to something like `type_hinting` or somehting like that?

Reply:
The compiler can't find the error type alone. I'm used to having the name very short because it generally appears much more often (dozens of times at least). It's only there once so I can use a turbofish instead